### PR TITLE
Add index to XPath information of XDocument equivalence failure message

### PR DESCRIPTION
**Pull request by lg2de**
I think, the XDocument equivalence failure message is not helpful when comparing documents with repeating element names, like HTML documents. By just providing the simple XPath it may be very difficult to identify the element with difference:
`/html/body/table/tr/td`

This PR is a proposal to improve the failure message by writing index number (according XPath syntax) when a specific element was found at least second time on same level:
`/html/body/table[2]/tr[17]/td[5]`

Additionally
* I recommend a new setting for `editorconfig` according to the current code style
* I changed a test name according to the test content

### CODE REVIEW

File: Src/FluentAssertions/Xml/XmlReaderValidator.cs [hunk 1]
```diff
@@ -77,7 +101,15 @@ private ValidationResult Validate()
                             return validationResult;
                         }
 
+                        // starting new element, add local name to location stack
+                        // to build XPath info
                         locationStack.Push(subjectReader.LocalName);
+
+                        // building key according XPath to count repeating element names
+                        var locationKey = "/" + string.Join("/", locationStack.Reverse());
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Xml/XmlReaderValidator.cs]: I don't get this. Aren't you building the entire path expression? How can that be used as a key?

File: Src/FluentAssertions/Xml/XmlReaderValidator.cs [hunk 2]
```diff
@@ -12,9 +13,32 @@ internal class XmlReaderValidator
         private readonly XmlReader subjectReader;
         private readonly XmlReader otherReader;
 
-        private string GetCurrentLocation() => "/" + string.Join("/", locationStack.Reverse());
+        private string GetCurrentLocation()
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Xml/XmlReaderValidator.cs]: I recommend renaming it to `GetCurrentXPath`. We also prefer to order methods by invocation so you can read from top to bottom. In this case, it means moving this method under `ValidateText`.

File: Src/FluentAssertions/Xml/XmlReaderValidator.cs [hunk 3]
```diff
@@ -12,9 +13,32 @@ internal class XmlReaderValidator
...
+                resultBuilder.Append($"/{location}");
+                var locationCount = locationCounts[keyBuilder.ToString()];
+                if (locationCount > 1)
+                {
+                    resultBuilder.Append($"[{locationCount}]");
+                }
+            }
+
+            string result = resultBuilder.ToString();
+            if (string.IsNullOrEmpty(result))
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Xml/XmlReaderValidator.cs]: This can never be `null`, right?

File: Src/FluentAssertions/Xml/XmlReaderValidator.cs [hunk 4]
```diff
@@ -12,9 +13,32 @@ internal class XmlReaderValidator
...
+            {
+                return "/";
+            }
+
+            return result;
+        }
 
         private readonly Stack<string> locationStack = new Stack<string>();
+        private readonly Dictionary<string, int> locationCounts = new Dictionary<string, int>();
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Xml/XmlReaderValidator.cs]: Isn't this the count per element? The current name confused me a bit.

File: Src/FluentAssertions/Xml/XmlReaderValidator.cs (hunk 1 above)
Comment (jnyrup) [reply Src/FluentAssertions/Xml/XmlReaderValidator.cs]: Given the xml `<a><b /><b /></a>`
* `<a>`: 
  * push `a` on the stack
  * update the dictionary with `/a` = 1
* `<b />`
  * push `b` on the stack
  * update the dictionary with `/a/b` = 1 <-- first time we see `<b>` on the `/a` xpath
  * pop `b` from the stack
* `<b />`
  * push `b` on the stack
  * update the dictionary with `/a/b` = 2 <-- second time we see `<b>` on the `/a` xpath
  * pop `b` from the stack
* `</a>`
  * pop `a` from the stack

File: Src/FluentAssertions/Xml/XmlReaderValidator.cs [hunk 5]
```diff
@@ -77,7 +101,15 @@ private ValidationResult Validate()
...
                         }
 
+                        // starting new element, add local name to location stack
+                        // to build XPath info
                         locationStack.Push(subjectReader.LocalName);
+
+                        // building key according XPath to count repeating element names
+                        var locationKey = "/" + string.Join("/", locationStack.Reverse());
+                        locationCounts.TryGetValue(locationKey, out var locationCount);
+                        locationCounts[locationKey] = ++locationCount;
```
Comment (jnyrup) [inline Src/FluentAssertions/Xml/XmlReaderValidator.cs]: 🔧How about extracting these lines into a method with some appropriate method name?

File: Src/FluentAssertions/Xml/XmlReaderValidator.cs (hunk 3 above)
Comment (jnyrup) [reply Src/FluentAssertions/Xml/XmlReaderValidator.cs]: Right, it cannot be `null`, at most empty. 

Could also be written as
```c#
if (resultBuilder.Length == 0)
{
    return "/";
}

string result = resultBuilder.ToString();
return result;
```

File: Src/FluentAssertions/Xml/XmlReaderValidator.cs (hunk 2 above)
Comment (lg2de) [reply Src/FluentAssertions/Xml/XmlReaderValidator.cs]: I will do so as soon you have merged #1174.

Comment (dennisdoomen) [reply Src/FluentAssertions/Xml/XmlReaderValidator.cs]: Done

**Comment by lg2de** [comment]
Do you think that this change of error message is a breaking change? 
If users asserting XPath using `WithMessage` existing tests may fail.

**Comment by jnyrup** [comment]
While it technically can be a breaking change, as in tests failing after upgrading the package, I don't recall we've considered improvements/changes to failure messages as breaking changes.

File: Tests/Shared.Specs/XDocumentAssertionSpecs.cs [hunk 6]
```diff
@@ -490,8 +490,23 @@ public void when_an_xml_element_has_attributes_and_is_self_closing_it_should_suc
...
+        [Fact]
+        public void When_asserting_equivalence_of_document_with_repeating_element_names_but_differs_it_should_fail_with_index_xpath_to_difference()
+        {
+            // Arrange
+            XDocument actual = XDocument.Parse(
+                "<xml><xml2 /><xml2 /><xml2><a x=\"y\"/><b><sub /></b><a x=\"y\"/></xml2></xml>");
+            XDocument expected = XDocument.Parse(
+                "<xml><xml2 /><xml2 /><xml2><a x=\"y\"/><b><sub /></b><a x=\"z\"/></xml2></xml>");
+
+            // Arrange
```
Comment (jnyrup) [inline Tests/Shared.Specs/XDocumentAssertionSpecs.cs]: `// Arrange` -> `// Act`

File: Src/FluentAssertions/Xml/XmlReaderValidator.cs (hunk 5 above)
Comment (lg2de) [reply Src/FluentAssertions/Xml/XmlReaderValidator.cs]: Mmh, this is the only usage, but it reduces the LOC in the `switch`.
Maybe, more statements should be moved to method (in another PR)!?

File: Tests/Shared.Specs/XDocumentAssertionSpecs.cs (hunk 6 above)
Comment (lg2de) [reply Tests/Shared.Specs/XDocumentAssertionSpecs.cs]: Nice catch!
Copied from existing test introduced by fa46521 from ... ;)

**Comment by dennisdoomen** [comment]
> While it technically can be a breaking change, as in tests failing after upgrading the package, I don't recall we've considered improvements/changes to failure messages as breaking changes.

Why would a test fail? Because somebody observe FA's failure messages in an exception assertion?

**Comment by lg2de** [comment]
> > While it technically can be a breaking change, as in tests failing after upgrading the package, I don't recall we've considered improvements/changes to failure messages as breaking changes.
> 
> Why would a test fail? Because somebody observe FA's failure messages in an exception assertion?

While creating an example I found that my question is not relevant.
Only when asserting FA failure message the change can cause failing tests. But this is not a normal use case.

**Comment by jnyrup** [comment]
I came to think of another way to combine the `Stack` and `Dictionary` into a single graph that keeps track of the xpath indices.

https://github.com/jnyrup/fluentassertions/commit/711a7f992a2cf0fb464d7180a76e36c1bab0afdf

Some very primitive benchmarks:
* `N` is the depth of the xml tree.
* xml tags are named `a0...aN`


# 10 children per node.

## This PR
```
|                     Method | N |         Mean |      Error |     StdDev | Ratio |    Gen 0 | Gen 1 | Gen 2 |  Allocated |
|--------------------------- |-- |-------------:|-----------:|-----------:|------:|---------:|------:|------:|-----------:|
| BeEquivalentTo             | 1 |     5.234 us |  0.1092 us |  0.1214 us |  1.00 |   1.3351 |     - |     - |    4.11 KB |
|                            |   |              |            |            |       |          |       |       |            |
| BeEquivalentTo             | 2 |    25.672 us |  0.2558 us |  0.2393 us |  1.00 |   4.6082 |     - |     - |   14.17 KB |
|                            |   |              |            |            |       |          |       |       |            |
| BeEquivalentTo             | 5 | 3,828.342 us | 30.7972 us | 24.0445 us |  1.00 | 648.4375 |     - |     - | 1995.18 KB |
```

## My branch
```
|                     Method | N |         Mean |      Error |     StdDev | Ratio |    Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------- |-- |-------------:|-----------:|-----------:|------:|---------:|------:|------:|----------:|
| BeEquivalentTo             | 1 |     2.987 us |  0.0256 us |  0.0227 us |  1.00 |   1.0529 |     - |     - |   3.24 KB |
|                            |   |              |            |            |       |          |       |       |           |
| BeEquivalentTo             | 2 |    12.788 us |  0.2393 us |  0.2122 us |  1.00 |   2.9602 |     - |     - |   9.14 KB |
|                            |   |              |            |            |       |          |       |       |           |
| BeEquivalentTo             | 5 | 1,558.388 us | 16.2188 us | 13.5434 us |  1.00 | 289.0625 |     - |     - | 890.32 KB |
```


# 4 children per node.

## This PR
```
|                     Method |  N |            Mean |         Error |        StdDev | Ratio |       Gen 0 | Gen 1 | Gen 2 |     Allocated |
|--------------------------- |--- |----------------:|--------------:|--------------:|------:|------------:|------:|------:|--------------:|
| BeEquivalentTo             |  3 |        71.90 us |      1.375 us |      1.351 us |  1.00 |     11.8408 |     - |     - |      36.59 KB |
|                            |    |                 |               |               |       |             |       |       |               |
| BeEquivalentTo             |  5 |     1,346.27 us |     36.026 us |     30.084 us |  1.00 |    224.6094 |     - |     - |     694.78 KB |
|                            |    |                 |               |               |       |             |       |       |               |
| BeEquivalentTo             | 10 | 1,871,360.79 us | 27,158.727 us | 24,075.504 us |  1.00 | 328000.0000 |     - |     - | 1009116.03 KB |
```

## My branch
```
|                     Method |  N |          Mean |         Error |       StdDev |        Median | Ratio |       Gen 0 | Gen 1 | Gen 2 |    Allocated |
|--------------------------- |--- |--------------:|--------------:|-------------:|--------------:|------:|------------:|------:|------:|-------------:|
| BeEquivalentTo             |  3 |      34.01 us |      0.678 us |     1.170 us |      33.47 us |  1.00 |      7.0190 |     - |     - |     21.63 KB |
|                            |    |               |               |              |               |       |             |       |       |              |
| BeEquivalentTo             |  5 |     534.87 us |      6.195 us |     5.492 us |     535.04 us |  1.00 |    101.5625 |     - |     - |    312.93 KB |
|                            |    |               |               |              |               |       |             |       |       |              |
| BeEquivalentTo             | 10 | 592,578.85 us | 10,082.794 us | 8,419.590 us | 594,102.50 us |  1.00 | 103000.0000 |     - |     - | 317702.62 KB |
```

**Comment by dennisdoomen** [comment]
@jnyrup do you want to do anything with those performance results? Or can this be merged?

**Comment by lg2de** [comment]
> @jnyrup do you want to do anything with those performance results? Or can this be merged?

Good question.

So far, I was unsure what to do.
I did not have time to analyzed in detail was @jnyrup has changed.
If performance is better, then I'm fine.
Should I merge it into "my" PR?

**Comment by jnyrup** [comment]
@lg2de You can merge the commit into this PR.
Its commit message (hopefully) describes the idea behind the visit count graph.
In the end it does the same as your algorithm, but more lazily to avoid stack iterations, string allocations and dictionary lookups.

> ### Use a counting graph to keep track of xpath index
> 
> Currently we use a stack to keep track of the location and a dictionary to keep track of the xpath index.
> That has multiple downsides:
> 
> * We need to keep two separate data structure in sync.
> * We're a lot of string allocations on each iterations when computing path as key for the dictionary.
> * When creating the failure message we're doing lookups in two data structures
> 
> Instead we can combine their purposes in a single graph where each node has a `Name` and a `Count`, where the latter is updated each time we visit a node with the same path.
> 
> On each xml traversal we now only need to find the matching sibling or insert a new child.
> 
> Creating the failure message is simply the travel from the current node to the root node in reverse order.

**Comment by lg2de** [comment]
@jnyrup why do you prefer
```
while (current.parent is object)
```
over
```
while (current.parent != null)
```
?

**Comment by dennisdoomen** [comment]
The second one can trigger an `!= operator`. The first one is a new C# construct that makes it immediately clear what you're expecting (providing you know the construct ;-))

**Comment by jnyrup** [comment]
https://intellitect.com/check-for-null-not-null/ has a nice write-up of different ways to compare against `null`.

By using `is object`, `is null` or `is {}` I don't have to care about whether a potentially overloaded `==` or `!=` operator has proper null checking.

Here's an example of the differences for value and reference types:
(The `{}` is a )
https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYYBGA9uQDboBqAhlQK4CmAKgJ4AOLcAFAEsAdgBd0AgJToAvAD5x4gM7pypAFYsAxiIDchYvqJp0FanUatOPGINHipchQEJp6IUypU9SYukMkTShoGZnZuFlRbMUkZeQElfABfb19/f2NTGgAlFgAzFgAnFiFNMJ5+XBjHeIFlVQ1tFIMfYgyg9Bz8opKylhtKhzj0FzcPLzSWozJ2zsLi0qsIvgHYhVqkpqJ/AHpt9AApJkUxEXJ0FgAPFgBbLioBXI50ehMmAHM3p4FbqhvikXoIgE5CE6QwUDgADZAmZyDwCoDyAVpNJls8ADQ4EyDdAiAAWBXIAHc3CwSQA5MazbqlACiF1KXCBIL4Ek2AQh0MyKnhiIKLjR9ExuFIOKcfBeKOx3kSQA===

**Comment by lg2de** [comment]
Mmh, understood. It feels overengineered to me.

I've applied jnyrup improvement (leaving the mentioned statement) and did some cleanup.
Would you like to merge now? Or do you still missing something?

**Comment by jnyrup** [comment]
Is that 
> leaving the mentioned statement

referring to this?
```
// starting new element, add local name to location stack
// to build XPath info
```

If so, everything looks good to me 👍

**Comment by lg2de** [comment]
No, sorry. 
I was referring to the null check. 
;)

**Comment by jnyrup** [comment]
@dennisdoomen whenever you're ready ;)

**Comment by lg2de** [comment]
> This class needs a bit of love. I'll take a look after merging this one.

What do you mean?

I just want to note, that I think about to provide new implementation for XDocument comparison.
Recently I worked often with XDocument/XElement and FluentAssertions. I found the output is still sometimes not helpful. I would like to have results like using `BeEquivalentTo` on complex classes. Hope I'll find time for this soon...

**Pull request merged**